x/trigger/types/genesis.go

```diff
@@ -0,0 +1,104 @@
+package types
+
+import (
+	fmt "fmt"
+
+	types "github.com/cosmos/cosmos-sdk/codec/types"
+	sdktx "github.com/cosmos/cosmos-sdk/types/tx"
+)
+
+func NewGenesisState(triggerID, queueStart uint64, triggers []Trigger, gasLimits []GasLimit, queuedTriggers []QueuedTrigger) *GenesisState {
+	return &GenesisState{
+		TriggerId:      triggerID,
+		QueueStart:     queueStart,
+		Triggers:       triggers,
+		GasLimits:      gasLimits,
+		QueuedTriggers: queuedTriggers,
+	}
+}
+
+// DefaultGenesis returns the default trigger genesis state
+func DefaultGenesis() *GenesisState {
+	return NewGenesisState(1, 1, []Trigger{}, []GasLimit{}, []QueuedTrigger{})
+}
+
+// Validate performs basic genesis state validation returning an error upon any
+// failure.
+func (gs GenesisState) Validate() error {
+	if gs.TriggerId == 0 {
+		return fmt.Errorf("invalid trigger id")
+	}
+	if gs.QueueStart == 0 {
+		return fmt.Errorf("invalid queue start")
+	}
+	if len(gs.Triggers)+len(gs.QueuedTriggers) != len(gs.GasLimits) {
+		return fmt.Errorf("gas limit list length must match sum of triggers and queued triggers length")
+	}
+
+	triggers := gs.Triggers
+	for _, queuedTrigger := range gs.QueuedTriggers {
+		triggers = append(triggers, queuedTrigger.GetTrigger())
+	}
```

Comment:
This can inadvertently cause problems.

Basically, if `gs.Triggers` has unused capacity, then as you append the queued triggers, they'd get added to that `GenesisState` field.

Basically, consider this situation:
* `gs.Triggers` has a length of 1 and capacity of 8 (this is common if it started as an empty slice or nil and the one entry was added using `append`.
* There are 7+ entries in `gs. QueuedTriggers`.

When you call `triggers := gs.Triggers` it copies the `gs.Triggers` variable to the new `triggers` variable. But since `gs.Triggers` is a slice, it's just a reference. So at that point, you have two references pointing to the same slice.

The first time you call `triggers = append(triggers, ...)`, the `append` function sees that there's 7 free spots in the slice, adds the new entry to the existing slice and updates it's length. The `append` function then returns the same reference that was provided since it didn't have to make a bigger slice.

The next 6 would end up doing the same thing. They'd get added to the original slice.

On the 8th one, `append` sees that the provided slice is out of space, so it creates a new, bigger underlying array copies the entries into the new one, and adds the entry to the new one. It then returns a different reference since it had to grow. So at this point, `triggers` and `gs.Triggers` are no longer the same reference.

And even though the receiver for this `Validate` method is concrete (e.g. it's `(gs GenesisState)` as opposed to `(gs *GenesisState)`), that only prevents the direct values in the struct from being altered. In this case, that value is just the reference to the slice. Since that reference isn't changing, the entries inadvertently added will remain at the end of the call to `Validate`.

There's a few ways to get around this, the simplest, code-wise is just `triggers := append([]Trigger{}, gs.Triggers...)`. There's arguments in goland about whether its better to do something like that or create a new slice (using `make(...)`) and then use `copy(dst, src)` to copy all the elements over. For our purposes, the performance difference is most likely negligible.

Reply:
This is a really good find, and makes perfect sense. Thanks for finding this.

For anyone else interested here is a working example: https://go.dev/play/p/kGu6J-geQm1